toolchain/sem_ir/generic.h

```diff
@@ -29,6 +31,74 @@ struct Generic : public Printable<Generic> {
   InstBlockId bindings_id;
 };
 
+// An instance of a generic entity, such as an instance of a generic function.
+// For each construct that depends on a compile-time parameter in the generic
+// entity, this contains the corresponding non-generic value. This includes
+// values for the compile-time parameters themselves.
+struct GenericInstance : Printable<GenericInstance> {
+  // Values corresponding to a region of a generic.
+  struct Region {
+    InstBlockId symbolic_constant_values_id = InstBlockId::Invalid;
+  };
+
+  auto Print(llvm::raw_ostream& out) const -> void {
+    out << "{generic: " << generic_id << ", args: " << args_id << "}";
+  }
+
+  // The generic that this is an instance of.
+  GenericId generic_id;
+  // Argument values, corresponding to the bindings in `Generic::bindings_id`.
+  InstBlockId args_id;
+};
+
+// Provides storage for deduplicated instances of generics.
+class GenericInstanceStore : public Yaml::Printable<GenericInstanceStore> {
+ public:
+  // Adds a new generic instance, or gets the existing generic instance for a
+  // specified generic and argument list. Returns the ID of the generic
+  // instance. The argument IDs must be for instructions in the constant block,
+  // and must be a canonical instruction block ID.
+  auto GetOrAdd(GenericId generic_id, InstBlockId args_id) -> GenericInstanceId;
+
+  // These are to support printable structures, and are not guaranteed.
+  auto OutputYaml() const -> Yaml::OutputMapping {
+    return generic_instances_.OutputYaml();
+  }
+
+ private:
+  // A lookup key for a generic instance.
+  struct Key {
+    GenericId generic_id;
+    InstBlockId args_id;
+
+    friend auto operator==(const Key&, const Key&) -> bool = default;
+  };
+
+  // Context for hashing keys.
+  struct KeyContext {
+    llvm::ArrayRef<GenericInstance> instances;
```

Comment:
Should this be private? It doesn't seem like it's part of the public API, so it strikes me as odd to put it first. hashtable_key_context.h's example provides the interface as a class, so would similar work here?

Reply:
This is just a struct; given the limited scope I don't think it makes sense to add encapsulation here.

Reply:
That said... looking at this again, I don't think this needs to be in the header file. Moved to the `.cpp` file.